src/hibernate-resume/hibernate-resume.c

```diff
@@ -1,58 +1,76 @@
-        }
-
         log_setup();
 
+        if (argc != 3)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "This program expects two arguments.");
+
         umask(0022);
 
-        /* Refuse to run unless we are in an initrd() */
         if (!in_initrd())
-                return EXIT_SUCCESS;
+                return 0;
 
-        device = argv[1];
+        arg_resume_device = argv[1];
 
-        if (stat(device, &st) < 0) {
-                log_error_errno(errno, "Failed to stat '%s': %m", device);
-                return EXIT_FAILURE;
-        }
+        r = safe_atou64(argv[2], &arg_resume_offset);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse resume offset %s: %m", argv[2]);
 
-        if (!S_ISBLK(st.st_mode)) {
-                log_error("Resume device '%s' is not a block device.", device);
-                return EXIT_FAILURE;
-        }
+        if (stat(arg_resume_device, &st) < 0)
+                return log_error_errno(errno, "Failed to stat resume device '%s': %m", arg_resume_device);
 
-        r = write_string_file("/sys/power/resume", FORMAT_DEVNUM(st.st_rdev), WRITE_STRING_FILE_DISABLE_BUFFER);
+        if (!S_ISBLK(st.st_mode))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Resume device '%s' is not a block device.", arg_resume_device);
+
+        r = write_string_filef("/sys/power/resume_offset", WRITE_STRING_FILE_DISABLE_BUFFER, "%" PRIu64, arg_resume_offset);
         if (r < 0) {
-                log_error_errno(r, "Failed to write '" DEVNUM_FORMAT_STR "' to /sys/power/resume: %m", DEVNUM_FORMAT_VAL(st.st_rdev));
-                return EXIT_FAILURE;
+                log_full_errno(r == -ENOENT && arg_resume_offset == 0 ? LOG_DEBUG : LOG_ERR,
+                               r,
+                               "Failed to write resume offset %" PRIu64 " to /sys/power/resume_offset for device '%s': %m",
+                               arg_resume_offset, arg_resume_device);
+                if (r != -ENOENT)
+                        return r;
         }
 
+        r = write_string_file("/sys/power/resume", FORMAT_DEVNUM(st.st_rdev), WRITE_STRING_FILE_DISABLE_BUFFER);
+        if (r < 0)
+                return log_error_errno(r, "Failed to write '" DEVNUM_FORMAT_STR "' to /sys/power/resume: %m", DEVNUM_FORMAT_VAL(st.st_rdev));
+
         /*
          * The write above shall not return.
          *
          * However, failed resume is a normal condition (may mean that there is
          * no hibernation image).
          */
 
-        log_info("Could not resume from '%s' (" DEVNUM_FORMAT_STR ").", device, DEVNUM_FORMAT_VAL(st.st_rdev));
-        return EXIT_SUCCESS;
+        log_info("Could not resume from device '%s' (" DEVNUM_FORMAT_STR ") offset %" PRIu64 ".",
+                 arg_resume_device,
+                 DEVNUM_FORMAT_VAL(st.st_rdev),
+                 arg_resume_offset);
```

Comment:
maybe reword to "Unable to resume from hibernation device …, continuing boot process."